### ISSUE TITLE
Indexing with negative step sizes

### ISSUE THREAD

**Issue by FilipeMaia** [arrayfire/arrayfire-python#24]
Currently arrays cannot be indexed with negative step sizes.

**Comment by FilipeMaia** [comment]
Do you want to support this? It doesn't seem so easy actually, so maybe raising an error is the best.

**Comment by FilipeMaia** [comment]
The C++ side can almost handle negative step sizes, the only obvious roadblock I see is there being no way to specify seq.begin at the end of the array. Using -1 triggers a [special code path](https://github.com/arrayfire/arrayfire/blob/devel/src/backend/dim4.cpp#L200) which I don't quite understand the purpose.

**Comment by FilipeMaia** [comment]
I think the main issue is that I don't quite understand how seq() is supposed to work for negative steps.

**Comment by pavanky** [comment]
Can this be closed after the PR ?

**Comment by FilipeMaia** [comment]
I'm afraid not, the issue is deeper. There's no easy way to do array[::-1] because there's no way (that I know) to specify that seq.begin corresponds to the end of the array in the C/C++ API. The fixes I did only handle cases where the begin and end are specified.

**Comment by pavanky** [comment]
af::seq(end,0, -1) does not work ?

**Comment by FilipeMaia** [comment]
Nope it gives:
```
af::exception: ArrayFire Exception(203): Invalid input size
In /Users/filipe/src/arrayfire/src/api/cpp/array.cpp:725
```

Code:
```
#include <arrayfire.h>

int main(){
  af::array a = af::randu(3);
  af_print(a);
  af_print(a(af::seq(af::end,0,-1)));
}
```

**Comment by pavanky** [comment]
OK, I think I can still handle this in python itself. If you are not working on this and #26 I will give a shot at fixing them.

**Comment by FilipeMaia** [comment]
If you look at calcDims when end is 0 you end up on:
 DIM_ASSERT(1, seq.begin >= -DBL_MIN && seq.begin < parentDim);

which will fail as af::end is smaller than -DBL_MIN

**Comment by FilipeMaia** [comment]
I'm not working on it, I have workarounds for my code.

**Comment by pavanky** [comment]
Ok, I think I am going to handle this upstream. The workarounds are ugly.

**Comment by pavanky** [comment]
This is fixed in 3.1.1 release. I think we can safely close this.
